### ISSUE TITLE
Add support for external Postgresql database

### ISSUE THREAD

**Issue by Frederickl78** [puppetlabs/puppetserver-helm-chart#170]
## Use Case
Add support for external Postgresql database like PAAS database to prevent managing stateful database in Kubernetes

## Describe the Solution You Would Like
A way to connect to external database.

## Describe Alternatives You've Considered
I have tried to work with current helm chart to configure external db connection; here's a diff of my value file config with original :
```
   postgresql:
     auth:
       username: puppetdb
       password: unbreakablePassword
       postgresPassword: unbreakableAdminPassword
       database: "puppet"

+     existingSecret: "puppetdb-postgresqlcredentials"     #<-- adding secret with username and password
...
    ## Extra initContainers to inject into the PuppetDB pod
    extraInitContainers: []
    ## Additional puppetdb container environment variables
    extraEnv:
+     PUPPETDB_POSTGRES_HOSTNAME: "postgres_hostname_fqdn"
+     PUPPETDB_POSTGRES_DATABASE: "puppetdbname"
```

And modifications to _helper.txt:
```
Return PostgreSQL host name
*/}}
...
 {{- printf "%s-%s" .Release.Name "postgresql-primary-hl" | trimSuffix "-" -}}
 {{- end -}}
+{{- else }}
+{{- if (hasKey .Values.puppetdb.extraEnv "PUPPETDB_POSTGRES_HOSTNAME") }}
+{{- printf .Values.puppetdb.extraEnv.PUPPETDB_POSTGRES_HOSTNAME }}
+{{- end -}}
 {{- end -}}
 {{- end -}}
```

Variable PUPPETDB_POSTGRES_HOSTNAME, PUPPETDB_POSTGRES_DATABASE, PUPPETDB_USER(user@postgresql_hostname) and PUPPETDB_PASSWORD in the pod are configured correctly but I have this error:
```
ERROR [p.p.c.services] Will retry database connection after temporary failure: java.sql.SQLTransientConnectionException: PDBMigrationsPool: default - Connection is not available, request timed out after 3000ms.
```

I tried to set variable PUPPETDB_POSTGRES_DATABASE with theses values with same error :
puppetdbname
puppetdbname?ssl=true&sslrootcert=/opt/puppetlabs/server/data/puppetdb/certs/certs/ca.pem
puppetdbname?ssl=true&sslfactory=org.postgresql.ssl.NonValidatingFactory

**Comment by Frederickl78** [comment]
Well; since nobody seems to look at bugs/enhancements...
I'm using azure postgresql Database so Im missing 2 parameters in database.conf.

So.  Configmap:
```
apiVersion: v1
data:
  database.conf: |
    database: {
      subname: "//"${PUPPETDB_POSTGRES_HOSTNAME}":"${PUPPETDB_POSTGRES_PORT}"/"${PUPPETDB_POSTGRES_DATABASE}
      username: ${PUPPETDB_USER}
      connection-username: ${PUPPETDB_CONNECTION_USERNAME}
      connection-migrator-username: ${PUPPETDB_CONNECTION_USERNAME}
      password: ${PUPPETDB_PASSWORD}
      node-ttl: ${PUPPETDB_NODE_TTL}
      node-purge-ttl: ${PUPPETDB_NODE_PURGE_TTL}
      report-ttl: ${PUPPETDB_REPORT_TTL}
    }
kind: ConfigMap
metadata:
  name: puppetdb-database-conf
```

Add database.conf volume and volumeMount to puppetdb-deployment.conf:
```
@@ -135,6 +135,9 @@ spec:
           volumeMounts:
             - name: puppetdb-storage
               mountPath: /opt/puppetlabs/server/data/puppetdb
+            - name: puppetdb-database
+              mountPath: /etc/puppetlabs/puppetdb/conf.d/database.conf
+              subPath: database.conf
             {{- if .Values.puppetdb.metrics.enabled }}
             - name: puppetdb-metrics-volume
               mountPath: /etc/puppetlabs/puppetdb/jolokia-access.xml
@@ -314,6 +317,9 @@ spec:
           configMap:
             name: puppetdb-custom-configs
         {{- end }}
+       - name: puppetdb-database
+          configMap:
+            name: puppetdb-database-conf
       {{- if .Values.nodeSelector }}
       nodeSelector:
         {{ toYaml .Values.nodeSelector | nindent 10 }}
```

Add PUPPETDB_CONNECTION_USERNAME extraEnv to puppetDB in value.yaml.
Well this is not ideal; This could be integrated in puppetdb docker image database.conf file.  Or maybe its a good thing to add the file as a configmap.

**Comment by davidphay** [comment]
Hi, I think you're request is linked to this issue: https://github.com/puppetlabs/puppetserver-helm-chart/issues/167 (except it's not for puppetserver but for puppetdb)

you want use an extra config map (in you're case to override an existing file)

And about this parts:
Return PostgreSQL host name

> */}}
> ...
>  {{- printf "%s-%s" .Release.Name "postgresql-primary-hl" | trimSuffix "-" -}}
>  {{- end -}}
> +{{- else }}
> +{{- if (hasKey .Values.puppetdb.extraEnv "PUPPETDB_POSTGRES_HOSTNAME") }}
> +{{- printf .Values.puppetdb.extraEnv.PUPPETDB_POSTGRES_HOSTNAME }}
> +{{- end -}}
>  {{- end -}}
>  {{- end -}}

It's not a good Idea to override postgres hostname like this:
Return PostgreSQL host name
*/}}
...
 {{- printf "%s-%s" .Release.Name "postgresql-primary-hl" | trimSuffix "-" -}}
 {{- end -}}
+{{- else }}
+{{- if ( eq  .Values.postgresql.enable false) and .Values.postgresql.hostname }}
+{{- printf .Values.postgresql.hostname }}
+{{- end -}}
 {{- end -}}
 {{- end -}}

In this case we can document the feature!

I don't have external postgresql so I can't write the feature, but please feel free to send our way a PR - I'll be happy to review and merge it, afterwards!

**Comment by nielshojen** [comment]
I've have added PR #175 with a change like the one you propose

I was getting the same error as you @Frederickl78  because I've forgotten to set PUPPETDB_USER and PUPPETDB_PASSWORD, Can you verify if you've set those

@davidphay I'm running with external an external db in google cloud, so can verify any changes

**Comment by davidphay** [comment]
Implemented in chart version 7.4.4
